### PR TITLE
[Inductor] FlexAttention supports partial masking

### PR DESCRIPTION
**Pull request by yanboliang**
This is the new version of #130235

Updated test script: https://gist.github.com/yanboliang/7c34a82df611d4ea8869cb9e041bfbfc
Updated perf numbers:
```
(pt) [ybliang@devgpu002.ash8 ~/local/debug]$ CUDA_VISIBLE_DEVICES=4 python debug7.py
fwd speedup: 0.7166695598192317
bwd speedup: 0.7142133867805904
(pt) [ybliang@devgpu002.ash8 ~/local/debug]$ CUDA_VISIBLE_DEVICES=4 python debug7.py --partial-mask
fwd speedup: 0.8428246087169973
bwd speedup: 0.8486261278030254
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_dynamo/variables/higher_order_ops.py [hunk 1]
```diff
@@ -1625,6 +1637,8 @@ def call_function(
             lse_meta = query_meta.new_empty(logsumexp_shape, dtype=torch.float32)
         example_value = (out_meta, lse_meta)
 
+        _, _, _, inp_arg_block_mask, inp_arg_scale = inp_args
+        block_mask = tuple(inp_arg_block_mask + (mask_fn_node,))
         # Compose the ordered HOO args from two parts:
         # - inp_args: [query, key, value, block_mask, scale]
         # - p_args: [score_mod, *other_buffers]
```
Comment (yanboliang) [inline torch/_dynamo/variables/higher_order_ops.py:1644]: I'll update comments later

File: torch/nn/attention/_flex_attention.py [hunk 2]
```diff
@@ -130,7 +173,9 @@ def sparsity(self) -> float:
         """
         total_size = self.numel()
         computed_size = (
-            self.kv_num_blocks.sum().item() * self.KV_BLOCK_SIZE * self.Q_BLOCK_SIZE
+            self.full_kv_num_blocks.sum().item()
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:176]: This should be full + partial blocks.

File: torch/_inductor/kernel/flex_attention.py [hunk 3]
```diff
@@ -528,18 +640,29 @@ def flex_attention(*args, **kwargs):
             # For now, we always assume the "sound" option
             ROWS_GUARANTEED_SAFE=False,
             PRESCALE_QK=False,
+            USE_PARTIAL_MASK=(sum(partial_kv_num_blocks.shape) != 3),
```
Comment (yanboliang) [inline torch/_inductor/kernel/flex_attention.py:643]: I think we have to figure out a better way to tell the Triton kernel if using partial mask, because I can't get if ```mask_fn``` is None from ```block_mask``` at here, probably we should add another flag.

File: torch/nn/attention/_flex_attention.py [hunk 4]
```diff
@@ -139,10 +184,12 @@ def to_dense(self) -> torch.Tensor:
         """
         Returns a dense block that is equivalent to the block mask.
         """
-        num_rows = self.kv_num_blocks.shape[-1]
-        num_cols = self.q_num_blocks.shape[-1]
-        batch_dims = self.kv_num_blocks.shape[:-1]
-        device = self.kv_num_blocks.device
+        num_rows = self.full_kv_num_blocks.shape[-1]
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:187]: I think this should also apply the mask_fn, but we can do that in a follow up.

File: torch/nn/attention/_flex_attention.py [hunk 5]
```diff
@@ -385,15 +526,18 @@ def create_block_mask(
         block_mask (tuple): A tuple of (kv_num_blocks, kv_indices, q_num_blocks, q_indices,
                             KV_BLOCK_SIZE, Q_BLOCK_SIZE) which represents the block mask.
     """
+    is_score_mod = len(inspect.signature(fn).parameters) == len(
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:529]: Unfortunately I don't think this works: 
image

File: torch/nn/attention/_flex_attention.py [hunk 6]
```diff
@@ -346,20 +433,74 @@ def _create_mask(
...
+    fn, B, H, M, N, device, KV_BLOCK_SIZE, Q_BLOCK_SIZE, is_score_mod
 ):
-    mask = _create_mask(score_mod, B, H, M, N, device, _compiled=True)
-    block_mask = _convert_mask_to_block_mask(
-        mask, KV_BLOCK_SIZE=KV_BLOCK_SIZE, Q_BLOCK_SIZE=Q_BLOCK_SIZE
+    if is_score_mod:
+        # fn is a score_mod function
+        mask_fn = None
+        mask_tensor = _create_mask_from_score_mod(
+            fn, B, H, M, N, device, _compiled=True
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:485]: I'm not sure this `_compiled` propagation logic is right... but I'm not sure it was right before either.

File: torch/nn/attention/_flex_attention.py [hunk 7]
```diff
@@ -278,37 +338,65 @@ def _convert_block_mask_to_mask(
...
+        )
+
     return BlockMask(
-        kv_num_blocks=kv_num_blocks.to(torch.int32).to(device).contiguous(),
-        kv_indices=kv_indices.to(torch.int32).to(device).contiguous(),
-        q_num_blocks=q_num_blocks.to(torch.int32).to(device).contiguous(),
-        q_indices=q_indices.to(torch.int32).to(device).contiguous(),
-        KV_BLOCK_SIZE=KV_BLOCK_SIZE,
-        Q_BLOCK_SIZE=Q_BLOCK_SIZE,
+        full_bm[0],
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:377]: You can simplify these by splatting them out I think, like `*full_bm, *partial_bm`.

File: torch/_dynamo/variables/higher_order_ops.py [hunk 8]
```diff
@@ -1602,12 +1605,21 @@ def call_function(
             scale,
         ) = self.normalize_to_args(args, kwargs)
 
-        p_args = self.create_wrapped_node(tx, query, score_mod)
+        score_mod_node, score_mod_lifted_args = self.create_wrapped_node(
+            tx, query, score_mod, "score_mod"
+        )
+        mask_fn = block_mask.items[-1]
+        if isinstance(mask_fn, ConstantVariable):
```
Comment (Chillee) [inline torch/_dynamo/variables/higher_order_ops.py:1612]: What does this do?

File: torch/_higher_order_ops/flex_attention.py [hunk 9]
```diff
@@ -137,17 +148,44 @@ def math_attention(
     m = torch.arange(0, scores.size(2), device=scores.device)
     n = torch.arange(0, scores.size(3), device=scores.device)
 
-    in_dim_buffers = (None,) * len(other_buffers)
-    score_mod = torch.vmap(score_mod, in_dims=(0, None, None, None, 0) + in_dim_buffers)
-    score_mod = torch.vmap(score_mod, in_dims=(0, None, None, 0, None) + in_dim_buffers)
-    score_mod = torch.vmap(score_mod, in_dims=(0, None, 0, None, None) + in_dim_buffers)
-    score_mod = torch.vmap(score_mod, in_dims=(0, 0, None, None, None) + in_dim_buffers)
+    score_mod_in_dim_buffers = (None,) * len(score_mod_other_buffers)
+    score_mod = torch.vmap(
```
Comment (Chillee) [inline torch/_higher_order_ops/flex_attention.py:152]: I feel like there's an utility function we could make here for all the use-cases of this pattern.

**Review by Chillee — COMMENTED** [review]
tentatively looks good :)

File: torch/_dynamo/variables/higher_order_ops.py (hunk 8 above)
Comment (yanboliang) [reply torch/_dynamo/variables/higher_order_ops.py:1612]: This is going to handle the case when mask_fn is None in eager mode.

File: torch/_higher_order_ops/flex_attention.py [hunk 10]
```diff
@@ -655,18 +778,28 @@ def trace_flex_attention_backward(
...
+    mask_example_vals = [torch.zeros((), dtype=torch.int) for _ in range(4)]
+    mask_graph = block_mask[-1]
     with TransformGetItemToIndex():
-        fw_graph = reenter_make_fx(fw_graph)(*fw_example_vals, *other_buffers)
-        joint_graph = reenter_make_fx(joint_graph)(*bw_example_vals, *other_buffers)
+        fw_graph = reenter_make_fx(fw_graph)(*fw_example_vals, *score_mod_other_buffers)
+        joint_graph = reenter_make_fx(joint_graph)(
+            *bw_example_vals, *score_mod_other_buffers
+        )
+        mask_graph = reenter_make_fx(mask_graph)(
```
[the remark this answers is not in the document]
Comment (yanboliang) [reply torch/_higher_order_ops/flex_attention.py:796]: Yes, this was not used, I can remove it.

File: test/inductor/test_flex_attention.py [hunk 11]
```diff
@@ -931,16 +933,16 @@ def f(q, k, v):
...
         fc = FileCheck()
-        fc.check("buf0 = empty_strided_cuda((1, 1, 1)")  # SPARSE_KV_NUM_BLKS
-        fc.check("buf1 = empty_strided_cuda((1, 1, 1, 1)")  # SPARSE_KV_IDX
-        fc.check("buf4 = empty_strided_cuda")  # logsumexp
-        fc.check("buf5 = empty_strided_cuda")  # attention output
-        fc.check("buf7 = empty_strided_cuda")  # cos(attention)
+        fc.check("buf0 = empty_strided_cuda((1, 1, 1)")  # FULL_KV_NUM_BLKS
+        fc.check("buf1 = empty_strided_cuda((1, 1, 1, 1)")  # FULL_KV_IDX
+        fc.check("buf8 = empty_strided_cuda")  # logsumexp
+        fc.check("buf11 = empty_strided_cuda")  # cos(attention)
```
Comment (peterbell10) [inline test/inductor/test_flex_attention.py:939]: The attention output still exists it's just called buf9 now.

File: torch/_inductor/select_algorithm.py [hunk 12]
```diff
@@ -86,10 +86,14 @@ def __init__(self, code, replacement_hooks):
         self.code = code
         self.replacement_hooks = replacement_hooks
 
-    def finalize_hook(self, hook_key: str) -> None:
-        assert (
-            hook_key in self.replacement_hooks
-        ), f"{hook_key} not registered in self.replacement_hooks"
+    def finalize_hook(self, hook_key: str, strict=True) -> None:
```
Comment (drisspg) [inline torch/_inductor/select_algorithm.py:89]: nit: I wish I knew why we need strict = False

File: torch/nn/attention/_flex_attention.py [hunk 13]
```diff
@@ -45,43 +52,74 @@ def _identity(
...
+
 _DEFAULT_SPARSE_BLOCK_SIZE = 128
 
 
 class BlockMask:
-    kv_num_blocks: torch.Tensor
-    kv_indices: torch.Tensor
-    q_num_blocks: torch.Tensor
-    q_indices: torch.Tensor
+    full_kv_num_blocks: torch.Tensor
```
Comment (drisspg) [inline torch/nn/attention/_flex_attention.py:68]: Nit: I think we should really add some documentation on this data structure should be interpreted. I know there is some by the kernel definition but I think it should be colocated here, or at least create a Note and reference it in relevant places

File: torch/nn/attention/_flex_attention.py [hunk 14]
```diff
@@ -45,43 +52,74 @@ def _identity(
...
+        mask_fn=None,
     ):
-        if kv_indices.dim() < 2:
-            raise RuntimeError("BlockMask kv_indices must have at least 2 dimensions")
-        self.kv_num_blocks = kv_num_blocks
-        self.kv_indices = kv_indices
-        self.q_num_blocks = q_num_blocks
-        self.q_indices = q_indices
+        if full_kv_indices.dim() < 2:
+            raise RuntimeError(
```
Comment (drisspg) [inline torch/nn/attention/_flex_attention.py:95]: do we need this for partial_kv_indices?

File: torch/nn/attention/_flex_attention.py [hunk 15]
```diff
@@ -92,25 +130,30 @@ def __str__(self):
         return s
 
     def __getitem__(self, index) -> "BlockMask":
-        tensors = self.as_tuple()[:-2]
+        tensors = self.as_tuple()[:-3]
         tensors = [x[index] for x in tensors]
         return BlockMask(
             tensors[0],
```
Comment (drisspg) [inline torch/nn/attention/_flex_attention.py:136]: ditto on splatting here

File: torch/_dynamo/variables/higher_order_ops.py (hunk 1 above)
Comment (yanboliang) [reply torch/_dynamo/variables/higher_order_ops.py]: Done

File: torch/_inductor/kernel/flex_attention.py (hunk 3 above)
Comment (yanboliang) [reply torch/_inductor/kernel/flex_attention.py]: I decided still to check if ```partial_kv_indices``` is ```torch.zeros([1,1,1])``` to mark if need partial masking inside of triton kernel.

File: torch/nn/attention/_flex_attention.py (hunk 2 above)
Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py]: Done

File: torch/nn/attention/_flex_attention.py (hunk 7 above)
Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py:396]: Tried, but it seems lint can't recognize it well, lol

File: torch/_higher_order_ops/flex_attention.py (hunk 9 above)
Comment (yanboliang) [reply torch/_higher_order_ops/flex_attention.py:139]: refactor out a ```math_attention_inner``` as both fwd and bwd would use the whole fwd pass.

File: torch/_higher_order_ops/flex_attention.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yanboliang) [reply torch/_higher_order_ops/flex_attention.py:790]: No, actually we can't, the bwd pass still need the fwd graph of ```mask_fn```, here we only trace though the fwd pass of mask fn.

File: test/inductor/test_flex_attention.py (hunk 11 above)
Comment (yanboliang) [reply test/inductor/test_flex_attention.py:940]: Done, thanks!

File: torch/nn/attention/_flex_attention.py (hunk 13 above)
Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py:68]: Yes, I think we need do a pass to fill all necessary docs and comments.

File: torch/nn/attention/_flex_attention.py (hunk 14 above)
Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py:95]: Done

File: torch/_inductor/select_algorithm.py (hunk 12 above)
Comment (Chillee) [reply torch/_inductor/select_algorithm.py:89]: Oh, it's that the other hooks are run on every single template. But in this case, not all templates have `gen_argdefs` on them.

**Review by Chillee — APPROVED** [review]
Tentative approve!

File: torch/_higher_order_ops/flex_attention.py (hunk 9 above)
Comment (Chillee) [reply torch/_higher_order_ops/flex_attention.py:139]: I meant to refactor out the `vmap` pattern, not the `math_attention_inner`.

**Comment by Chillee** [comment]
Landed as https://github.com/pytorch/pytorch/pull/130626
